torch/distributed/_composable/checkpoint_activation.py

```diff
@@ -214,63 +62,39 @@ def checkpoint(module: nn.Module, *, use_reentrant: bool = True) -> nn.Module:
-                # SavedVariable data is cleared.
-                storage: WeakKeyDictionary = WeakKeyDictionary()
-                weak_holder_list: List[ReferenceType] = []
-                saved_tensor_hooks = torch.autograd.graph.saved_tensors_hooks(
-                    partial(_pack, weak_holder_list=weak_holder_list),
-                    partial(
-                        _unpack,
-                        storage=storage,
-                        weak_holder_list=weak_holder_list,
-                        module=module,
-                        inputs=inputs,
-                    ),
-                )
-                saved_tensor_hooks.__enter__()
-                checkpoint.state(module).saved_tensor_hooks = saved_tensor_hooks
+            def context_fns():
+                return nullcontext(), _no_hook(module)
+
+            checkpoint.state(
+                module
+            )._ac_generator = _checkpoint_without_reentrant_generator(
+                module,
+                True,
+                context_fns,
+                _DEFAULT_DETERMINISM_MODE,
+                False,
+                *inputs
+            )
+            next(checkpoint.state(module)._ac_generator)
 
     def forward_hook(module: nn.Module, inputs: Tuple[Any, ...], output: Any) -> Any:
         if checkpoint.state(module).enable_hook:
-            torch.set_grad_enabled(checkpoint.state(module).orig_grad_enabled)
-            if checkpoint.state(module).use_reentrant:
-                return _ModuleHookCheckpointFunction.apply(module, output, *inputs)
+            try:
+                next(checkpoint.state(module)._ac_generator)
+            except StopIteration:
+                pass
             else:
-                checkpoint.state(module).saved_tensor_hooks.__exit__()
-                checkpoint.state(module).saved_tensor_hooks = None
+                raise RuntimeError(
+                    "Expected non-reentrant activation checkpoint generator to be exhausted, but it was not!"
+                )
+
+        #  Ensure that we no longer hold on to the generator. always_call=True helps ensure we
+        # clear this even in the case of exception in fwd pass.
+        checkpoint.state(module)._ac_generator = None
 
-        return output
 
-    # This hook does the following things:
-    # 1. detach outputs from the autograd graph to discard activations
-    # 2. insert an autograd.Function after the forward pass to recompute
-    #    activations during the backward pass.
     checkpoint.state(module).enable_hook = True
-    checkpoint.state(module).use_reentrant = use_reentrant
     module.register_forward_pre_hook(forward_pre_hook)
-    # Use prepend to make sure we restore the original grad enabled state right
-    # after the module forward invocation.
-    module.register_forward_hook(forward_hook, prepend=True)
+    module.register_forward_hook(forward_hook, prepend=True, always_call=True)
```

Comment:
Given that we are not passing kwargs into the hooks, I think we will still need to fix for supporting kwargs.

Reply:
sounds good, we can fix this in the follow up diffs?